README.md
# **`ipzone`**

Ipzone provides a simple and powerful IP architecture to Rust.

[![CI][ci-badge]][ci-url]
[![Crates.io][crates-badge]][crates-url]
[![Licensed][license-badge]][license-url]
[![Twitter][twitter-badge]][twitter-url]

[ci-badge]: https://github.com/just-do-halee/ipzone/actions/workflows/ci.yml/badge.svg
[crates-badge]: https://img.shields.io/crates/v/ipzone.svg?labelColor=383636
[license-badge]: https://img.shields.io/crates/l/ipzone?labelColor=383636
[twitter-badge]: https://img.shields.io/twitter/follow/do_halee?style=flat&logo=twitter&color=4a4646&labelColor=333131&label=just-do-halee
[ci-url]: https://github.com/just-do-halee/ipzone/actions
[twitter-url]: https://twitter.com/do_halee
[crates-url]: https://crates.io/crates/ipzone
[license-url]: https://github.com/just-do-halee/ipzone

| [Examples](./examples/) | [Docs](https://docs.rs/ipzone) | [Latest Note](./CHANGELOG.md) |

```toml
ipzone = "0.4.0"
```

## **`Examples`**

```rust
use ipzone::prelude::*;

let local: Address<3> = ip::localhost([6004, 7040, 8080]);
TcpListener::bind(local);

let address: Address<2> = ip::from([168, 159, 42, 9]).with([80, 443]);
TcpStream::connect(address);

let local = ip::localhost([port::from_env("PORT").unwrap_or(8080), 7020, 2020]);
TcpListener::bind(local);

let address = ip::from([186, 23, 123, 1, 0, 0, 0, 0]).with([80, 443]);
TcpStream::connect(address);

let local = ip::localhost([1234, 5678, port::from_str("9090").unwrap());
TcpListener::bind(local);

let address = ip::from_str("168.24.41.123").unwrap().with([80, 443]);
TcpStream::connect(address);

let local = ip::from_str("::1").unwrap().with([80, 443]);
TcpListener::bind(local);

let address = ip::from_env("IP").unwrap().with([80, 443]);
TcpStream::connect(address);

static LOCALHOST: Global<Address> = global(|| ip::localhost().with([8080]));
```



CHANGELOG.md
<!-- next-header -->

## [0.4.0] - 2022-12-10

## Improved

- `#[repr(transparent)]` for Port(pub u16)

Released by [@just-do-halee](https://github.com/just-do-halee).

<!-- next-url -->

[0.4.0]: https://github.com/just-do-halee/stdto/compare/v0.4.0...HEAD


tests/ip_from.rs
use ipzone::prelude::*;
use std::env;

#[test]
fn ipv4_array() {
    let ip = ip::from([186, 23, 123, 1]).with([80, 443]);
    assert_eq!(
        ip.to_socket_addrs().unwrap().collect::<Vec<_>>(),
        vec![
            SocketAddr::new(Ipv4::new(186, 23, 123, 1).into(), 80),
            SocketAddr::new(Ipv4::new(186, 23, 123, 1).into(), 443),
        ]
    );
}

#[test]
fn ipv6_array() {
    let ip = ip::from([186, 23, 123, 1, 0, 0, 0, 0]).with([80, 443]);
    assert_eq!(
        ip.to_socket_addrs().unwrap().collect::<Vec<_>>(),
        vec![
            SocketAddr::new(Ipv6::new(186, 23, 123, 1, 0, 0, 0, 0).into(), 80),
            SocketAddr::new(Ipv6::new(186, 23, 123, 1, 0, 0, 0, 0).into(), 443),
        ]
    );
}

#[test]
fn ipv4_u32() {
    let ip = ip::from(0x0a000001u32.to_be_bytes()).with([80, 443]);
    assert_eq!(
        ip.to_socket_addrs().unwrap().collect::<Vec<_>>(),
        vec![
            SocketAddr::new(Ipv4::new(10, 0, 0, 1).into(), 80),
            SocketAddr::new(Ipv4::new(10, 0, 0, 1).into(), 443),
        ]
    );
}

#[test]
fn ipv6_u128() {
    let ip = ip::from(0x00000000000000000000ffffaabbccddu128.to_be_bytes()).with([80, 443]);
    assert_eq!(
        ip.to_socket_addrs().unwrap().collect::<Vec<_>>(),
        vec![
            SocketAddr::new(Ipv6::new(0, 0, 0, 0, 0, 0xffff, 0xaabb, 0xccdd).into(), 80),
            SocketAddr::new(Ipv6::new(0, 0, 0, 0, 0, 0xffff, 0xaabb, 0xccdd).into(), 443),
        ]
    );
}

#[test]
fn ipv4_str() {
    let ip = ip::from_str("164.23.1.2").unwrap().with([80, 443]);
    assert_eq!(
        ip.to_socket_addrs().unwrap().collect::<Vec<_>>(),
        vec![
            SocketAddr::new(Ipv4::new(164, 23, 1, 2).into(), 80),
            SocketAddr::new(Ipv4::new(164, 23, 1, 2).into(), 443),
        ]
    );
}

#[test]
fn ipv6_str() {
    let ip = ip::from_str("::ffff:aabb:ccdd").unwrap().with([80, 443]);
    assert_eq!(
        ip.to_socket_addrs().unwrap().collect::<Vec<_>>(),
        vec![
            SocketAddr::new(Ipv6::new(0, 0, 0, 0, 0, 0xffff, 0xaabb, 0xccdd).into(), 80),
            SocketAddr::new(Ipv6::new(0, 0, 0, 0, 0, 0xffff, 0xaabb, 0xccdd).into(), 443),
        ]
    );
}

#[test]
fn ipv4_env() {
    env::set_var("IP", "127.0.0.1");
    let ip = ip::from_env("IP").unwrap().with([80, 443]);
    assert_eq!(
        ip.to_socket_addrs().unwrap().collect::<Vec<_>>(),
        vec![
            SocketAddr::new(Ipv4::new(127, 0, 0, 1).into(), 80),
            SocketAddr::new(Ipv4::new(127, 0, 0, 1).into(), 443),
        ]
    );
}

#[test]
fn ipv6_env() {
    env::set_var("IP", "::ffff:aabb:ccdd");
    let ip = ip::from_env("IP").unwrap().with([80, 443]);
    assert_eq!(
        ip.to_socket_addrs().unwrap().collect::<Vec<_>>(),
        vec![
            SocketAddr::new(Ipv6::new(0, 0, 0, 0, 0, 0xffff, 0xaabb, 0xccdd).into(), 80),
            SocketAddr::new(Ipv6::new(0, 0, 0, 0, 0, 0xffff, 0xaabb, 0xccdd).into(), 443),
        ]
    );
}


tests/localhost.rs
use ipzone::prelude::*;

#[test]
fn it_works() {
    let localhost = ip::localhost().with([6060, 8080]);

    assert_eq!(
        localhost.to_socket_addrs().unwrap().collect::<Vec<_>>(),
        vec![
            SocketAddr::new(Ip::V4(Ipv4::new(127, 0, 0, 1)), 6060),
            SocketAddr::new(Ip::V4(Ipv4::new(127, 0, 0, 1)), 8080),
        ]
    );
}


tests/port_from.rs
use ipzone::prelude::*;
use std::env;

#[test]
fn it_works() {
    let localhost = ip::localhost().with([6060, 8080, 1234, 5678]);
    assert_eq!(
        localhost.to_socket_addrs().unwrap().collect::<Vec<_>>(),
        vec![
            SocketAddr::new(Ip::V4(Ipv4::new(127, 0, 0, 1)), 6060),
            SocketAddr::new(Ip::V4(Ipv4::new(127, 0, 0, 1)), 8080),
            SocketAddr::new(Ip::V4(Ipv4::new(127, 0, 0, 1)), 1234),
            SocketAddr::new(Ip::V4(Ipv4::new(127, 0, 0, 1)), 5678),
        ]
    );
}

#[test]
fn it_works_for_str() {
    let ip = ip::from([186, 23, 123, 1]).with(["80".parse().unwrap(), 443, 123, 5678]);
    assert_eq!(
        ip.to_socket_addrs().unwrap().collect::<Vec<_>>(),
        vec![
            SocketAddr::new(Ipv4::new(186, 23, 123, 1).into(), 80),
            SocketAddr::new(Ipv4::new(186, 23, 123, 1).into(), 443),
            SocketAddr::new(Ipv4::new(186, 23, 123, 1).into(), 123),
            SocketAddr::new(Ipv4::new(186, 23, 123, 1).into(), 5678),
        ]
    );
}

#[test]
fn port_from_str() {
    let ip = ip::localhost().with([
        //
        80,
        port::from_str("4240").unwrap(),
        port::from_str("asdasd").unwrap_or(2020),
    ]);
    assert_eq!(
        ip.to_socket_addrs().unwrap().collect::<Vec<_>>(),
        vec![
            SocketAddr::new(Ip::V4(Ipv4::new(127, 0, 0, 1)), 80),
            SocketAddr::new(Ip::V4(Ipv4::new(127, 0, 0, 1)), 4240),
            SocketAddr::new(Ip::V4(Ipv4::new(127, 0, 0, 1)), 2020),
        ]
    );
}

#[test]
fn port_from_env() {
    env::set_var("PORT", "8080");
    let ip = ip::localhost().with([
        //
        80,
        port::from_env("PORT").unwrap(),
    ]);
    assert_eq!(
        ip.to_socket_addrs().unwrap().collect::<Vec<_>>(),
        vec![
            SocketAddr::new(Ip::V4(Ipv4::new(127, 0, 0, 1)), 80),
            SocketAddr::new(Ip::V4(Ipv4::new(127, 0, 0, 1)), 8080),
        ]
    );
}


src/lib.rs
#![cfg(not(doctest))]
#![doc = include_str!("../README.md")]

#[macro_use]
mod global;
mod address;

pub mod ip;
pub mod port;

pub mod prelude {
    pub use super::{ip, port};

    pub use super::{address::*, cmn::FromEnvError, global::*};

    pub use std::net::IpAddr as Ip;
    pub use std::net::Ipv4Addr as Ipv4;
    pub use std::net::Ipv6Addr as Ipv6;
    pub use std::net::SocketAddr;
    pub use std::net::ToSocketAddrs;
}

// -----------
//
mod cmn;
use cmn::*;

use port::*;
use prelude::*;

#[cfg(test)]
mod tests {
    use super::*;

    use std::net::{Ipv6Addr, SocketAddr};

    #[test]
    fn it_works() {
        std::env::set_var("IP", "::ffff:192.1.56.10");
        std::env::set_var("PORT", "8080");

        let address = ip::from_env("IP").unwrap().with([
            80,
            port::from_str("123").unwrap(),
            port::from_env("PORT").unwrap(),
        ]);

        assert_eq!(
            address.to_socket_addrs().unwrap().collect::<Vec<_>>(),
            vec![
                SocketAddr::new(
                    Ipv6Addr::new(0, 0, 0, 0, 0, 0xffff, 0xc001, 0x380a).into(),
                    80
                ),
                SocketAddr::new(
                    Ipv6Addr::new(0, 0, 0, 0, 0, 0xffff, 0xc001, 0x380a).into(),
                    123
                ),
                SocketAddr::new(
                    Ipv6Addr::new(0, 0, 0, 0, 0, 0xffff, 0xc001, 0x380a).into(),
                    8080
                ),
            ]
        );
    }
}


src/port.rs
use super::*;
use std::{num::ParseIntError, str::FromStr};

#[inline]
pub fn from(source: impl Into<u16>) -> Port {
    Port(source.into())
}
#[inline]
pub fn from_str(s: &str) -> Result<u16, ParseIntError> {
    _from_str(s)
}
#[inline]
pub fn from_env(key: &str) -> Result<u16, FromEnvError> {
    _from_env(key)
}

#[repr(transparent)]
#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct Port(pub u16);

impl From<Port> for u16 {
    #[inline]
    fn from(port: Port) -> Self {
        port.0
    }
}

impl From<u16> for Port {
    #[inline]
    fn from(port: u16) -> Self {
        Self(port)
    }
}
impl TryFrom<&str> for Port {
    type Error = ParseIntError;

    #[inline]
    fn try_from(s: &str) -> Result<Self, Self::Error> {
        Ok(Self(_from_str(s)?))
    }
}
impl FromStr for Port {
    type Err = ParseIntError;
    #[inline]
    fn from_str(s: &str) -> Result<Self, Self::Err> {
        Ok(Self(s.parse()?))
    }
}


src/address.rs
use super::*;
use std::{array, io};

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct Address<const N: usize = 1>(pub Ip, pub [Port; N]);

impl<const N: usize> Address<N> {
    #[inline]
    pub fn first(&self) -> SocketAddr {
        SocketAddr::new(self.0, self.1[0].into())
    }
    #[inline]
    pub fn last(&self) -> SocketAddr {
        SocketAddr::new(self.0, self.1[N - 1].into())
    }
    #[inline]
    pub fn iter(&self) -> array::IntoIter<SocketAddr, N> {
        self.into_iter()
    }
}

impl<const N: usize> IntoIterator for Address<N> {
    type Item = SocketAddr;
    type IntoIter = array::IntoIter<SocketAddr, N>;

    #[inline]
    fn into_iter(self) -> Self::IntoIter {
        self.1
            .map(move |port| SocketAddr::new(self.0, port.0))
            .into_iter()
    }
}

impl<const N: usize> ToSocketAddrs for Address<N> {
    type Iter = array::IntoIter<SocketAddr, N>;
    #[inline]
    fn to_socket_addrs(&self) -> io::Result<Self::Iter> {
        Ok(self.into_iter())
    }
}

pub trait Dock {
    fn to_ip(&self) -> Ip;
    #[inline]
    fn with<const N: usize>(&self, ports: [impl Into<Port>; N]) -> Address<N> {
        let ports = ports.map(|port| port.into());
        Address(self.to_ip(), ports)
    }
}

impl Dock for Ip {
    #[inline]
    fn to_ip(&self) -> Ip {
        *self
    }
}


examples/simple.rs
use std::net::{IpAddr, Ipv4Addr, Ipv6Addr, SocketAddr};

use ipzone::prelude::*;

fn main() {
    std::env::set_var("PORT", "8080");

    let localhost = ip::localhost().with([
        //
        1234,
        port::from_env("PORT").unwrap(),
        port::from_env("PORT2").unwrap_or(5050),
    ]);
    assert_eq!(
        localhost.to_socket_addrs().unwrap().collect::<Vec<_>>(),
        vec![
            SocketAddr::new(IpAddr::V4(Ipv4Addr::new(127, 0, 0, 1)), 1234),
            SocketAddr::new(IpAddr::V4(Ipv4Addr::new(127, 0, 0, 1)), 8080),
            SocketAddr::new(IpAddr::V4(Ipv4Addr::new(127, 0, 0, 1)), 5050),
        ]
    );

    let ip = ip::from([186, 23, 123, 1]).with([
        //
        80,
        port::from_str("123").unwrap(),
        port::from_env("PORT").unwrap(),
    ]);
    assert_eq!(
        ip.to_socket_addrs().unwrap().collect::<Vec<_>>(),
        vec![
            SocketAddr::new(Ipv4Addr::new(186, 23, 123, 1).into(), 80),
            SocketAddr::new(Ipv4Addr::new(186, 23, 123, 1).into(), 123),
            SocketAddr::new(Ipv4Addr::new(186, 23, 123, 1).into(), 8080),
        ]
    );

    let ip2 = ip::from_str("::ffff:aabb:ccdd").unwrap().with([80, 443]);
    assert_eq!(
        ip2.to_socket_addrs().unwrap().collect::<Vec<_>>(),
        vec![
            SocketAddr::new(
                Ipv6Addr::new(0, 0, 0, 0, 0, 0xffff, 0xaabb, 0xccdd).into(),
                80
            ),
            SocketAddr::new(
                Ipv6Addr::new(0, 0, 0, 0, 0, 0xffff, 0xaabb, 0xccdd).into(),
                443
            ),
        ]
    );

    std::env::set_var("IP", "168.32.12.1");
    let ip3 = ip::from_env("IP").unwrap().with([80, 443]);
    assert_eq!(
        ip3.to_socket_addrs().unwrap().collect::<Vec<_>>(),
        vec![
            SocketAddr::new(Ipv4Addr::new(168, 32, 12, 1).into(), 80),
            SocketAddr::new(Ipv4Addr::new(168, 32, 12, 1).into(), 443),
        ]
    );
}
